File: src/nac/basic/source_range.cpp
#include "sodium/nac/basic/source_range.h"

#include "sodium/nac/basic/source_location.h"

namespace sodium {

SourceRange::SourceRange(SourceLocation start, SourceLocation end) : start_(start), end_(end) {}

SourceLocation SourceRange::start() const {
    return start_;
}

SourceLocation SourceRange::end() const {
    return end_;
}

} // namespace sodium

File: nac/src/basic/source_range.rs
use std::fmt::Display;

use super::source_location::SourceLocation;

pub struct SourceRange {
    start: SourceLocation,
    end: SourceLocation,
}

impl SourceRange {
    pub fn new(start: SourceLocation, end: SourceLocation) -> Self {
        Self { start, end }
    }

    pub fn start(&self) -> &SourceLocation {
        &self.start
    }

    pub fn end(&self) -> &SourceLocation {
        &self.end
    }
}

impl Display for SourceRange {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} -> {}", self.start(), self.end())
    }
}
